pkg/apis/v1beta1/aksnodeclass.go

```diff
@@ -96,6 +104,128 @@ type Security struct {
 	EncryptionAtHost *bool `json:"encryptionAtHost,omitempty"`
 }
 
+// +kubebuilder:validation:Enum:={Preferred,Required,Disabled}
+type LocalDNSMode string
+
+const (
+	// If the current orchestrator version supports this feature, prefer enabling localDNS.
+	LocalDNSModePreferred LocalDNSMode = "Preferred"
+	// Enable localDNS.
+	LocalDNSModeRequired LocalDNSMode = "Required"
+	// Disable localDNS.
+	LocalDNSModeDisabled LocalDNSMode = "Disabled"
+)
+
+// LocalDNS configures the per-node local DNS, with VnetDNS and KubeDNS overrides.
+// LocalDNS helps improve performance and reliability of DNS resolution in an AKS cluster.
+// For more details see aka.ms/aks/localdns.
+// +kubebuilder:validation:XValidation:rule="'.' in self.vnetDNSOverrides && 'cluster.local' in self.vnetDNSOverrides",message="vnetDNSOverrides must contain required zones '.' and 'cluster.local'"
+// +kubebuilder:validation:XValidation:rule="'.' in self.kubeDNSOverrides && 'cluster.local' in self.kubeDNSOverrides",message="kubeDNSOverrides must contain required zones '.' and 'cluster.local'"
+// +kubebuilder:validation:XValidation:rule="!('.' in self.vnetDNSOverrides && has(self.vnetDNSOverrides['.'].forwardDestination) && self.vnetDNSOverrides['.'].forwardDestination == 'ClusterCoreDNS')",message="DNS traffic for root zone '.' cannot be forwarded to ClusterCoreDNS from vnetDNSOverrides"
+// +kubebuilder:validation:XValidation:rule="!self.vnetDNSOverrides.exists(zone, zone.endsWith('cluster.local') && has(self.vnetDNSOverrides[zone].forwardDestination) && self.vnetDNSOverrides[zone].forwardDestination == 'VnetDNS')",message="DNS traffic for 'cluster.local' cannot be forwarded to VnetDNS from vnetDNSOverrides"
+// +kubebuilder:validation:XValidation:rule="!self.kubeDNSOverrides.exists(zone, zone.endsWith('cluster.local') && has(self.kubeDNSOverrides[zone].forwardDestination) && self.kubeDNSOverrides[zone].forwardDestination == 'VnetDNS')",message="DNS traffic for 'cluster.local' cannot be forwarded to VnetDNS from kubeDNSOverrides"
+type LocalDNS struct {
+	// Mode of enablement for localDNS.
+	// +required
+	Mode *LocalDNSMode `json:"mode"`
+	// VnetDNS overrides apply to DNS traffic from pods with dnsPolicy:default or kubelet (referred to as VnetDNS traffic).
+	// +required
+	VnetDNSOverrides map[string]*LocalDNSOverrides `json:"vnetDNSOverrides"`
```

Comment:
nomaps: field LocalDNS.VnetDNSOverrides should not use a map type, use a list type with a unique name/identifier instead (kubeapilinter)